tls/cfsslscan_handshake.go

```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
 	if err != nil {
 		return
 	}
+	if CipherSuites[serverHello.cipherSuite].EllipticCurve {
+		// Prime the connection, if necessary, for key
+		// exchange messages by reading off the certificate
+		// message and, if necessary, the OCSP stapling
+		// message
+		var msg interface{}
+		msg, err = c.readHandshake()
+		if err != nil {
+			return
+		}
+		certMsg, ok := msg.(*certificateMsg)
+		if !ok || len(certMsg.certificates) == 0 {
+			c.sendAlert(alertUnexpectedMessage)
+			err = unexpectedMessageError(certMsg, msg)
+			return
+		}
+		if serverHello.ocspStapling {
+			msg, err = c.readHandshake()
+			if err != nil {
+				return
+			}
+			_, ok := msg.(*certificateStatusMsg)
+			if !ok {
+				c.sendAlert(alertUnexpectedMessage)
+				return
+			}
+		}
+
+		var skx *serverKeyExchangeMsg
+		skx, err = c.exchangeKeys()
+		if skx.raw[0] != typeServerKeyExchange {
+			err = errors.New("tls: received something other than the expected ServerKeyExchange.")
+			return
+		}
+		if err != nil {
+			return
+		}
+		if len(skx.key) < 4 {
+			err = errServerKeyExchange
+			return
+		}
+		curveType = uint16(skx.key[0])
+		// If we have a named curve, report which one it is.
+		if curveType == 3 {
+			curveID = CurveID(skx.key[1])<<8 | CurveID(skx.key[2])
+		}
+	}
 	cipherID, version = serverHello.cipherSuite, serverHello.vers
 	return
-
 }
 
 // sayHello is the backend to SayHello that returns a full serverHelloMsg for processing.
```

Comment:
Which sites cause this error? It might be caused by a bug in some F5 load balancers (https://www.imperialviolet.org/2014/02/27/tlssymmetriccrypto.html).

> But it was the F5 workaround that caused most of the headaches. Older (but common) firmware revisions of F5 devices hang the connection when the ClientHello is longer than 255 bytes. This appeared to be a major issue because it meant that we couldn't add things to the ClientHello without breaking lots of big sites that use F5 devices. After a plea and a discussion on the TLS list, an engineer from F5 explained that it was just handshake messages between 256 and 511 bytes in length that caused the issue. That suggested an obvious solution: pad the handshake message so that it doesn't fall into the troublesome size range.